prime-router/src/main/kotlin/ActionLog.kt

```diff
@@ -146,8 +68,149 @@ class ActionError(val details: List<ActionLog>, message: String? = null) : Error
             var message = super.message ?: ""
             if (details.isNotEmpty()) {
                 message += System.lineSeparator() +
-                    details.joinToString(System.lineSeparator()) { it.detail.detailMsg() }
+                    details.joinToString(System.lineSeparator()) { it.detail.message }
             }
             return message
         }
+}
+
+/**
+ * Details for a given action log.
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "class")
+@JsonIgnoreProperties(ignoreUnknown = true)
+interface ActionLogDetail {
+    /**
+     * The scopt of the log.
+     */
+    val scope: ActionLogScope
+
+    /**
+     * The log message.
+     */
+    val message: String
+}
+
+/**
+ * A logger for action logs.
+ */
+class ActionLogger() {
+    /**
+     * The raw logs.
+     */
+    val logs = mutableListOf<ActionLog>()
+
+    /**
+     * The current item index being tracked, or null if no index is tracked.
+     */
+    private var itemIndex: Int? = null
+
+    /**
+     * The current item tracking ID being tracked, or null if no tracking ID is tracked.
+     */
+    private var trackingId: String? = null
+
+    /**
+     * The report ID of the logs if available.
+     */
+    private var reportId: UUID? = null
+
+    /**
+     * Start item logging for a given [itemIndex] and optional [trackingId].
+     * @return the logger instance
+     */
+    fun startItemLogging(itemIndex: Int, trackingId: String? = null) = apply {
+        check(itemIndex > 0) { "Item index must be a positive number" }
+        this.itemIndex = itemIndex
+        this.trackingId = trackingId
+    }
```

Comment:
The problem here is that a lot of the code does not know the item number or tracking ID, therefore it is needed at a different level.  Either that or a major refactor.

Reply:
There should be a couple of ways to make it more atomic. In the places this is used it could be replaced by calls like `actionLogs.error(errors, trackingId, itemIndex)`.

Another option is to do something like
```
ActionLogger(
  val logs: MutableList<ActionLog>=mutableListOf<ActionLog>(),
  val trackingId: String? = null,
  val itemIndex: Int? = null,
  etc..
)
//... and then when you have a trackingId
val itemLevelLogger = ActionLogger(actionLogs.logs, trackingId, itemIndex)
```

Reply:
I like this last option.  Let me play with it!